File: src/services/permissionService.ts
import { OrganizationRole, Project, ProjectMembership, ProjectRole, User } from '../types';

export interface Permissions {
  canEditProject: boolean;
  isProjectAdmin: boolean;
  isOrgAdmin: boolean;
}

export const getProjectMembership = (
  userId: string,
  projectId: string,
  projectMemberships: ProjectMembership[]
): ProjectMembership | undefined => {
  return projectMemberships.find(pm => pm.userId === userId && pm.projectId === projectId);
};

export const getUserOrganizationRole = (
  userId: string,
  organizationId: string,
  users: User[]
): OrganizationRole | undefined => {
  const user = users.find(u => u.id === userId && u.organizationId === organizationId);
  return user?.role;
};

export const calculatePermissions = (
  userId: string,
  project: Project | null,
  users: User[],
  projMemberships: ProjectMembership[]
): Permissions => {
  if (!project) {
    // Default permissions when no project is selected
    const currentUser = users.find(u => u.id === userId);
    const isAnyOrgAdmin = currentUser?.role === OrganizationRole.ORG_ADMIN;
    return { canEditProject: false, isProjectAdmin: false, isOrgAdmin: isAnyOrgAdmin || false };
  }

  const userRole = getUserOrganizationRole(userId, project.organizationId, users);
  const projectMembership = getProjectMembership(userId, project.id, projMemberships);

  const isOrgAdmin = userRole === OrganizationRole.ORG_ADMIN;
  const isProjectAdmin = projectMembership?.role === ProjectRole.PROJECT_ADMIN;

  const canEdit = isOrgAdmin || isProjectAdmin || projectMembership?.role === ProjectRole.DESIGNER;

  return {
    canEditProject: canEdit,
    isProjectAdmin: isProjectAdmin || isOrgAdmin,
    isOrgAdmin: isOrgAdmin,
  };
};

File: backend/Services/PermissionService.cs
using TaraFusion.Data;
using TaraFusion.DTOs;
using TaraFusion.Models;
using Microsoft.EntityFrameworkCore;

namespace TaraFusion.Services;

public class PermissionService
{
    private readonly ApplicationDbContext _context;

    public PermissionService(ApplicationDbContext context)
    {
        _context = context;
    }

    public async Task<ProjectMembership?> GetProjectMembership(Guid userId, Guid projectId)
    {
        return await _context.ProjectMemberships
            .FirstOrDefaultAsync(pm => pm.UserId == userId && pm.ProjectId == projectId);
    }

    public async Task<OrganizationMembership?> GetOrganizationMembership(Guid userId, Guid organizationId)
    {
        return await _context.OrganizationMemberships
            .FirstOrDefaultAsync(om => om.UserId == userId && om.OrganizationId == organizationId);
    }

    public async Task<PermissionsDto> CalculatePermissions(Guid userId, Guid projectId)
    {
        var project = await _context.Projects.FindAsync(projectId);
        if (project == null)
        {
            return new PermissionsDto { CanEditProject = false, IsProjectAdmin = false, IsOrgAdmin = false };
        }

        var orgMembership = await GetOrganizationMembership(userId, project.OrganizationId);
        var projectMembership = await GetProjectMembership(userId, projectId);

        var isOrgAdmin = orgMembership?.Role == "ORG_ADMIN";
        var isProjectAdmin = projectMembership?.Role == "PROJECT_ADMIN";

        var canEdit = isOrgAdmin || isProjectAdmin || projectMembership?.Role == "DESIGNER";

        return new PermissionsDto
        {
            CanEditProject = canEdit,
            IsProjectAdmin = isProjectAdmin || isOrgAdmin,
            IsOrgAdmin = isOrgAdmin,
        };
    }
}
